### DISCUSSION TITLE
Feature request: Ability to order vehicles

### DISCUSSION THREAD

**Discussion by leewillis77** [teslamate-org/teslamate#1751] [Ideas]
**Describe the feature**

When you have multiple vehicles in a single Teslamate instance, the main Teslamate web interface shows the vehicles in "order added". For example if you acquire vehicles called "B", then "A", then "C" then that is the order you'll see them in (top-to-bottom) on the main Teslamate screen. 

On the grafana reports though the cars are displayed in alphabetical order, so you see car "A", then "B", then "C" . 

Rather than just fixing that inconsistency though, it would be great if we could set a "priority", or "order" for vehicles. In my case, the car that I'm interested in 90% of the time always appears second due to naming, and it would be super helpful if I could change those to appear in my chosen order.

**Expected behavior**

Ability to set an explicit order for all vehicles, or ability to set a "priority" value against each vehicle on the Teslamate settings page, and have both the teslamate web screen, and Grafana dashboards honour it.

**Comment by adriankumpf** [comment]
Feature requests belong here: [💡 Ideas](https://github.com/adriankumpf/teslamate/discussions/categories/ideas). I'll move it.

**Comment by tobrz** [comment]
I'd love this as well, take my donation money for it :-)

**Comment by DrMichael** [comment]
> On the grafana reports though the cars are displayed in alphabetical order, so you see car "A", then "B", then "C" .

I think it is displayed in order of the car id in the cars table. Options would be to introduce a prio in cars, but then you also would need an interface to manage that. Or to give the user the ability to change the car id...

**Reply by leewillis77** [reply]
> I think it is displayed in order of the car id in the cars table.

Definitely not. My car displays last in grafana, despite being the first car added to our install.

**Reply by DrMichael** [reply]
What gives `select name, id from cars;` in the database container?
(Hard to test with only one car... :-)

**Reply by leewillis77** [reply]
```
# SELECT id, name, inserted_at, updated_at FROM cars;
 id |           name           |     inserted_at     |     updated_at      
----+--------------------------+---------------------+---------------------
  2 | Annie                    | 2020-09-23 15:32:32 | 2021-06-09 12:49:07
  1 | Garlic Bread             | 2020-08-19 17:09:41 | 2021-05-29 10:22:03
```

**Reply by DrMichael** [reply]
And which order have the rows?:-O

**Reply by leewillis77** [reply]
OK, so it's not doing explicit name ordering. Instead the order is "undefined" which is even worse ;)

**Reply by DrMichael** [reply]
Hard to tell with two entries...

Could you try defining car_id as `SELECT name AS __text, id AS __value FROM cars ORDER BY id ASC;` and one with `DESC` at the end?

**Comment by leewillis77** [comment]
I think this could be implemented incrementally. 

Step 1
------
Add "priority"/"order" to the cars table
Amend grafana reports to order by priority/name/car ID

Step 2
-------
Implement UI for setting order

I keep meaning to look at this, but I don't have a dev env setup so I need enough time to take a good run at it. It's definitely on my radar as I think just implementing step 1 should be fairly straightforward.

**Reply by DrMichael** [reply]
Perhaps try it out:

> ## Step 1
> Add "priority"/"order" to the cars table

Add order to the cars table in the database container.

> Amend grafana reports to order by priority/name/car ID

Define car_id as `SELECT name AS __text, order AS __value FROM cars;`

**Reply by leewillis77** [reply]
Oh, I can implement it pretty easily on my own install, but it's getting a dev env setup so I can build the DB migrations etc. that's the time consuming bit :)

**Reply by DrMichael** [reply]
Ah, adding a column would not harm an existing installation... Neither would a new (private) dashboard.

**Comment by leewillis77** [comment]
Started work on this here: https://github.com/adriankumpf/teslamate/pull/1904/
